src/main/scala/firrtl/transforms/InlineBooleanExpressions.scala

```diff
@@ -0,0 +1,187 @@
+  // private because it's not clear if this definition of "Simple Expression" would be useful elsewhere.
+  // Note that this can have false negatives but MUST NOT have false positives.
+  def isSimpleExpr(expr: Expression): Boolean = expr match {
+    case _: WRef | _: Literal | _: WSubField => true
+    case DoPrim(op, args, _,_) if isBooleanExpr(op) => args.forall(isSimpleExpr)
+    case _ => false
+  }
+
+  /** Mapping from references to the [[firrtl.ir.Expression Expression]]s that drive them */
+  type Netlist = mutable.HashMap[WrappedExpression, Expression]
+
+  /** Recursively replace [[WRef]]s with new [[Expression]]s
+    *
+    * @param netlist a '''mutable''' HashMap mapping references to [[firrtl.ir.DefNode DefNode]]s to their connected
+    * [[firrtl.ir.Expression Expression]]s. It is '''not''' mutated in this function
+    * @param expr the Expression being transformed
+    * @return Returns expr with Bits inlined
+    */
+  def onExpr(netlist: Netlist, symbolTable: Map[WrappedExpression, Int], depth: Int)(expr: Expression): Expression = {
+    expr.map(onExpr(netlist, symbolTable, depth + 1)) match {
+      case lhs @ DoPrim(lop, args, lc, ltpe) if isSimpleExpr(lhs) =>
+        val ivalNew = args.map {
+          e =>
+            val we1 = we(e)
+            val rhs = netlist.get(we1)
+            rhs match {
+              case Some(e1) if (isBooleanExpr(e1)) =>
+                if (symbolTable(we1) <= 1) { // if the fanout of the expression is 1 then replace
+                  e1
+                }
+                else {
+                  e
+                }
+              case _ => e
+            }
+        }
+        val l = lhs.copy(args = ivalNew)
+        l
+      case other => other // Not a candidate
+    }
+  }
+
+  /**
+    * During the recursive traversal of the rhs expression the depth that the boolean expressions are combined is controlled
+    * StartingDepth is set to 0
+    */
+  val StartingDepth = 0
+
+  /** Inline bits in a Statement
+    *
+    * @param netlist a '''mutable''' HashMap mapping references to [[firrtl.ir.DefNode DefNode]]s to their connected
+    * [[firrtl.ir.Expression Expression]]s. This function '''will''' mutate it if stmt is a [[firrtl.ir.DefNode
+    * DefNode]] with a Temporary name and a value that is a [[PrimOp]] Bits
+    * @param stmt the Statement being searched for nodes and transformed
+    * @return Returns stmt with Bits inlined
+    */
+  def onStmt(netlist: Netlist, symbolTable: Map[WrappedExpression, Int])(stmt: Statement): Statement =
+  stmt.map(onStmt(netlist, symbolTable)).map(onExpr(netlist, symbolTable, StartingDepth))
+
+  /** Replaces bits in a Module */
+  def onMod(mod: DefModule,
+    netlist: Netlist,
+    symbolTable: Map[WrappedExpression, Int]): DefModule =
+    mod.map(onStmt(netlist, symbolTable))
+}
+
+/** Inline nodes that are simple bits */
+class InlineBooleanExpressions extends Transform {
+  def inputForm = LowForm
+  def outputForm = LowForm
```

Comment:
@azidar can we go over the From's at the weekly meeting next week?
In particular UnknownForm.

Reply:
Not sure how to fix this problem:

Exception in thread "main" java.lang.AssertionError: assertion failed: No transform in List(firrtl.ChirrtlToHighFirrtl@1184ab05, firrtl.IRToWorkingIR@3aefe5e5, firrtl.ResolveAndCheck@149e0f5d, firrtl.transforms.DedupModules@1b1473ab, firrtl.HighFirrtlToMiddleFirrtl@2f7c2f4f, firrtl.MiddleFirrtlToLowFirrtl@6af93788, firrtl.LowFirrtlOptimization@ef9296d) has outputForm UnknownForm as required by firrtl.transforms.InlineBooleanExpressions@7d3a22a9
	at scala.Predef$.assert(Predef.scala:223)
	at firrtl.CompilerUtils$.$anonfun$mergeTransforms$2(Compiler.scala:385)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:89)
	at firrtl.CompilerUtils$.mergeTransforms(Compiler.scala:382)
	at firrtl.Compiler.compile(Compiler.scala:475)
	at firrtl.Compiler.compile$(Compiler.scala:474)
	at firrtl.VerilogCompiler.compile(LoweringCompilers.scala:167)
	at firrtl.stage.phases.Compiler.f$1(Compiler.scala:91)
	at firrtl.stage.phases.Compiler.$anonfun$internalTransform$2(Compiler.scala:95)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at firrtl.stage.phases.Compiler.internalTransform(Compiler.scala:95)
	at firrtl.stage.phases.Compiler.internalTransform(Compiler.scala:45)
	at firrtl.options.Translator.transform(Phase.scala:140)
	at firrtl.options.Translator.transform$(Phase.scala:140)
	at firrtl.stage.phases.Compiler.transform(Compiler.scala:45)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:32)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:15)
	at firrtl.options.Translator.transform(Phase.scala:140)
	at firrtl.options.Translator.transform$(Phase.scala:140)
	at firrtl.options.phases.DeletedWrapper.transform(DeletedWrapper.scala:15)
	at firrtl.stage.FirrtlStage.$anonfun$run$1(FirrtlStage.scala:27)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:89)
	at firrtl.stage.FirrtlStage.run(FirrtlStage.scala:27)
	at firrtl.options.Stage$$anon$1.transform(Stage.scala:43)
	at firrtl.options.Stage$$anon$1.transform(Stage.scala:43)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:32)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:15)
	at firrtl.options.Translator.transform(Phase.scala:140)
	at firrtl.options.Translator.transform$(Phase.scala:140)
	at firrtl.options.phases.DeletedWrapper.transform(DeletedWrapper.scala:15)
	at firrtl.options.Stage.$anonfun$transform$5(Stage.scala:46)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:89)
	at firrtl.options.Stage.$anonfun$transform$3(Stage.scala:46)
	at logger.Logger$.$anonfun$makeScope$2(Logger.scala:168)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at logger.Logger$.makeScope(Logger.scala:166)
	at firrtl.options.Stage.transform(Stage.scala:46)
	at firrtl.options.Stage.transform(Stage.scala:17)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:32)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:15)
	at firrtl.options.Translator.transform(Phase.scala:140)
	at firrtl.options.Translator.transform$(Phase.scala:140)
	at firrtl.options.phases.DeletedWrapper.transform(DeletedWrapper.scala:15)
	at firrtl.Driver$.$anonfun$execute$2(Driver.scala:222)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:89)
	at firrtl.Driver$.execute(Driver.scala:222)
	at firrtl.Driver$.execute(Driver.scala:241)
	at firrtl.Driver$.main(Driver.scala:257)
	at firrtl.Driver.main(Driver.scala)

Reply:
@seldridge Thanks for your PR.

Reply:
Couple of things going on here:

1) Once something is `UnknownForm`/`UnknownForm`, it can no longer be scheduled using `--custom-transforms` (which is what I think is the stack trace above).
2) Any `*TransformSpec` for a transform that's `UnknownForm`/`UnknownForm` can no longer be used to build tests. (This is a deficiency in the testing infrastructure that needs to be rectified once the Dependency API, a means of specifying dependencies between transforms, lands and deprecates `CircuitForm`.)

I think this means that the testing for `InlineBooleanExpressions` would be better to integrate with the `VerilogEmitterTests` directly (e.g., similar PRs in this vein have worked that way like https://github.com/freechipsproject/firrtl/pull/1296/files#diff-a5a5a2065905cbe57265344603f513ba).

All that said, it's not the end of the world to leave this as `LowForm`/`LowForm`. I'm just trying to avoid that if it's possible and reasonable. Probably worth asking @azidar.

Reply:
@seldridge Here is a test case which replicates the bug shown in your RocketCore trace. The symbolTable (reference counter might be a better name) does not contain `sel` because only RHS references in Connect and DefNode statements are added to the symbolTable. What is needed is to add all uses of the sel in conditions and select as well to the symbol table. Is there a firrtl `shortcut` like the foreach you used elsewhere which will capture all uses of `sel`?

```
circuit Top :
  module Top :
    input clk : Clock
    input a : UInt<1>
    input b : UInt<1>
    input s0 : UInt<1>
    input s1 : UInt<1>
    output x: UInt<1>
    node sel = and(s0, s1)
    x <= mux(sel, a, b)
    printf(clk, not(sel), "xxxx\n")
```

Reply:
Nice minimum working example! I think you can simplify this further with just the `printf` and show that tracking across `printf` needs to work, too.

I'm not aware of a utility method that builds up this netlist for you, but it may exist (you may ping @azidar and see if he knows of one). But, the limitation of `onExpr` only tracking simple `DoPrim`s would need to be removed here.